File: backend/src/ai/behavioral_analyzer.js
class BehavioralAnalyzer {
  constructor() {
    console.log('🧠 Behavioral Analyzer AI initialized');
  }

  async analyze(userText, userHistory) {
    const sentiment = this.analyzeSentiment(userText);
    const patterns = this.analyzePatterns(userHistory);
    const recommendations = this.generateRecommendations(sentiment, patterns);
    
    return {
      sentiment,
      patterns,
      recommendations,
      confidence: this.calculateConfidence(userHistory)
    };
  }

  analyzeSentiment(text) {
    const positiveWords = ['happy', 'excited', 'confident', 'motivated', 'success'];
    const negativeWords = ['worried', 'stressed', 'anxious', 'frustrated', 'overwhelmed'];
    
    const lowerText = text.toLowerCase();
    let positiveCount = 0;
    let negativeCount = 0;
    
    positiveWords.forEach(word => {
      if (lowerText.includes(word)) positiveCount++;
    });
    
    negativeWords.forEach(word => {
      if (lowerText.includes(word)) negativeCount++;
    });
    
    if (positiveCount > negativeCount) return 'positive';
    if (negativeCount > positiveCount) return 'negative';
    return 'neutral';
  }

  analyzePatterns(history) {
    if (!history || history.length === 0) {
      return { consistency: 'unknown', trend: 'stable', risk: 'low' };
    }
    
    const contributions = history.filter(h => h.type === 'contribution');
    const consistency = this.calculateConsistency(contributions);
    const trend = this.calculateTrend(contributions);
    const risk = this.assessRisk(consistency, trend);
    
    return { consistency, trend, risk };
  }

  calculateConsistency(contributions) {
    if (contributions.length < 2) return 'unknown';
    
    const intervals = [];
    for (let i = 1; i < contributions.length; i++) {
      const days = (new Date(contributions[i].date) - new Date(contributions[i-1].date)) / (1000 * 60 * 60 * 24);
      intervals.push(days);
    }
    
    const avgInterval = intervals.reduce((a, b) => a + b, 0) / intervals.length;
    const variance = intervals.reduce((a, b) => a + Math.pow(b - avgInterval, 2), 0) / intervals.length;
    
    if (variance < 7) return 'high';
    if (variance < 14) return 'medium';
    return 'low';
  }

  calculateTrend(contributions) {
    if (contributions.length < 3) return 'stable';
    
    const amounts = contributions.slice(-3).map(c => c.amount);
    const firstHalf = amounts.slice(0, Math.ceil(amounts.length / 2));
    const secondHalf = amounts.slice(Math.ceil(amounts.length / 2));
    
    const firstAvg = firstHalf.reduce((a, b) => a + b, 0) / firstHalf.length;
    const secondAvg = secondHalf.reduce((a, b) => a + b, 0) / secondHalf.length;
    
    if (secondAvg > firstAvg * 1.1) return 'increasing';
    if (secondAvg < firstAvg * 0.9) return 'decreasing';
    return 'stable';
  }

  assessRisk(consistency, trend) {
    if (consistency === 'low' && trend === 'decreasing') return 'high';
    if (consistency === 'low' || trend === 'decreasing') return 'medium';
    return 'low';
  }

  generateRecommendations(sentiment, patterns) {
    const recommendations = [];
    
    if (sentiment === 'negative') {
      recommendations.push('Consider setting smaller, more achievable goals to build confidence');
    }
    
    if (patterns.consistency === 'low') {
      recommendations.push('Set up automatic contributions to improve consistency');
    }
    
    if (patterns.trend === 'decreasing') {
      recommendations.push('Review your budget and identify areas to reduce expenses');
    }
    
    if (patterns.risk === 'high') {
      recommendations.push('Focus on building an emergency fund before pursuing larger goals');
    }
    
    return recommendations;
  }

  calculateConfidence(history) {
    if (!history || history.length === 0) return 0.5;
    return Math.min(0.5 + (history.length * 0.1), 0.95);
  }
}

module.exports = { BehavioralAnalyzer }; 

File: backend/ai/behavioral_analyzer.py
import json
import re
from datetime import datetime, timedelta

class BehavioralAnalyzer:
    def __init__(self):
        """Initialize the AI Behavioral Analyzer"""
        print("✅ AI Behavioral Analyzer initialized successfully!")
    
    def analyze(self, userText, userHistory):
        """Analyze user behavior patterns and provide insights"""
        try:
            # Analyze spending patterns from text
            spending_insights = self._analyzeSpendingPatterns(userText)
            
            # Analyze savings behavior
            savings_insights = self._analyzeSavingsBehavior(userHistory)
            
            # Generate behavioral recommendations
            recommendations = self._generateRecommendations(spending_insights, savings_insights)
            
            return {
                "success": True,
                "analysis": {
                    "spending_patterns": spending_insights,
                    "savings_behavior": savings_insights,
                    "recommendations": recommendations,
                    "timestamp": datetime.now().isoformat()
                }
            }
            
        except Exception as e:
            return {
                "success": False,
                "error": f"Error analyzing behavior: {str(e)}"
            }
    
    def _analyzeSpendingPatterns(self, userText):
        """Analyze spending patterns from user text"""
        insights = []
        
        # Look for spending-related keywords
        spending_keywords = ['spend', 'bought', 'purchase', 'expense', 'cost', 'price', 'shopping']
        saving_keywords = ['save', 'budget', 'cut', 'reduce', 'limit']
        
        text_lower = userText.lower()
        
        # Analyze spending vs saving language
        spending_count = sum(1 for word in spending_keywords if word in text_lower)
        saving_count = sum(1 for word in saving_keywords if word in text_lower)
        
        if spending_count > saving_count:
            insights.append("💸 Your language suggests a spending-focused mindset")
            insights.append("💡 Consider reframing goals in terms of what you're saving for")
        elif saving_count > spending_count:
            insights.append("💰 Your language shows a savings-focused mindset")
            insights.append("🌟 Great job maintaining this perspective!")
        else:
            insights.append("⚖️ Balanced approach to spending and saving")
        
        # Look for emotional spending indicators
        emotional_words = ['stress', 'bored', 'sad', 'excited', 'impulse', 'treat']
        emotional_count = sum(1 for word in emotional_words if word in text_lower)
        
        if emotional_count > 0:
            insights.append("😊 Be mindful of emotional spending triggers")
            insights.append("💡 Try the 24-hour rule for non-essential purchases")
        
        # Look for budget awareness
        if any(word in text_lower for word in ['budget', 'plan', 'track']):
            insights.append("📊 You're showing good budget awareness")
        else:
            insights.append("📝 Consider tracking your spending to identify patterns")
        
        return insights
    
    def _analyzeSavingsBehavior(self, userHistory):
        """Analyze savings behavior from user history"""
        insights = []
        
        if not userHistory or len(userHistory) == 0:
            insights.append("🆕 Welcome! Let's start building your savings habits")
            return insights
        
        # Analyze contribution frequency
        recent_contributions = [h for h in userHistory if h.get('type') == 'contribution']
        
        if len(recent_contributions) >= 3:
            insights.append("🎯 Consistent savings behavior detected")
            insights.append("💪 You're building great financial habits")
        elif len(recent_contributions) >= 1:
            insights.append("👍 Good start with savings")
            insights.append("🔄 Try to make savings a regular habit")
        else:
            insights.append("💡 Consider setting up automatic savings transfers")
        
        # Analyze goal progress
        goals = [h for h in userHistory if h.get('type') == 'goal']
        if goals:
            active_goals = [g for g in goals if g.get('status') == 'active']
            completed_goals = [g for g in goals if g.get('status') == 'completed']
            
            if completed_goals:
                insights.append("🏆 You've successfully completed financial goals")
                insights.append("🌟 Celebrate your achievements!")
            
            if active_goals:
                insights.append(f"🎯 You have {len(active_goals)} active savings goals")
                insights.append("📈 Keep pushing toward your targets")
        
        return insights
    
    def _generateRecommendations(self, spending_insights, savings_insights):
        """Generate personalized behavioral recommendations"""
        recommendations = []
        
        # Spending recommendations
        if any("spending-focused" in insight for insight in spending_insights):
            recommendations.append("💡 Try the 50/30/20 rule: 50% needs, 30% wants, 20% savings")
            recommendations.append("📱 Use LoopFund to track your spending categories")
        
        if any("emotional" in insight for insight in spending_insights):
            recommendations.append("🧘 Practice mindful spending - pause before purchases")
            recommendations.append("📝 Keep a spending journal to identify triggers")
        
        # Savings recommendations
        if any("consistent" in insight for insight in savings_insights):
            recommendations.append("🚀 Consider increasing your monthly savings amount")
            recommendations.append("🎯 Set more ambitious financial goals")
        
        if any("automatic" in insight for insight in savings_insights):
            recommendations.append("⚡ Set up automatic transfers on payday")
            recommendations.append("🎉 Make saving invisible and automatic")
        
        # General recommendations
        recommendations.append("📊 Review your progress weekly to stay motivated")
        recommendations.append("🎯 Break large goals into smaller milestones")
        recommendations.append("💪 Remember: consistency beats perfection")
        
        return recommendations
